### PR TITLE
Fix/symmetric contraction groupmm bindings

### PR DESCRIPTION
**Pull request by asglover**
Re-enable symmetric contraction / groupmm bindings and added basic pytest testing.

### CODE REVIEW

File: openequivariance/openequivariance/_torch/extlib/__init__.py [hunk 1]
```diff
@@ -89,15 +97,17 @@ def load_jit_extension():
 
             extra_cflags.append("-DCUDA_BACKEND")
         elif torch.version.hip:
-            extra_link_args.extend(["-lhiprtc"])
+            extra_link_args.extend(["-lhiprtc", "-lrocblas"])
```
Comment (vbharadwaj-bk) [inline openequivariance/openequivariance/_torch/extlib/__init__.py]: Hmm.. have we tested this? This looks reasonable but I'm wary about changing this without a platform to test on.

File: openequivariance/openequivariance/_torch/extlib/__init__.py [hunk 2]
```diff
@@ -65,6 +65,14 @@ def load_jit_extension():
...
 
         include_dirs, extra_link_args = (["backend"], ["-Wl,--no-as-needed"])
+        extra_include_dirs = []
+
+        try:
+            import pybind11
+
+            extra_include_dirs.append(pybind11.get_include())
+        except Exception as e:
+            getLogger().info(f"Could not locate pybind11 include path: {e}")
```
Comment (vbharadwaj-bk) [inline openequivariance/openequivariance/_torch/extlib/__init__.py]: Given that PyBind11 is required for JIT extension compilation, you should set BUILT_EXTENSION_ERROR here rather than log the error.

File: openequivariance/openequivariance/_torch/extlib/__init__.py (hunk 1 above)
Comment (vbharadwaj-bk) [reply openequivariance/openequivariance/_torch/extlib/__init__.py]: It should be fine if this code was in our main branch at some point in the past - I excised a lot of this symmetric contraction at some point.

File: openequivariance/CMakeLists.txt [hunk 3]
```diff
@@ -117,6 +117,7 @@ endfunction()
 
 find_package(CUDAToolkit QUIET)
 find_package(hip QUIET)
+find_package(rocblas QUIET)
```
Comment (vbharadwaj-bk) [inline openequivariance/CMakeLists.txt:120]: Ditto, checking here that we had this in our codebase in the past, removed it, and we are putting it back. If so, no problem.

File: openequivariance/CMakeLists.txt [hunk 4]
```diff
@@ -157,13 +159,20 @@ if(hip_FOUND)
         CXX_STANDARD 17
     )
 
+    if(TARGET roc::rocblas)
```
Comment (vbharadwaj-bk) [inline openequivariance/CMakeLists.txt:162]: Hmm... have we tested this? Combo of RocBLAS + stable extension, I have not verified.

File: openequivariance/CMakeLists.txt [hunk 5]
```diff
@@ -137,6 +138,7 @@ if(CUDAToolkit_FOUND)
         CUDA::cudart
         CUDA::cuda_driver
         CUDA::nvrtc
+        CUDA::cublas
```
Comment (vbharadwaj-bk) [inline openequivariance/CMakeLists.txt:141]: This is fine because we're addressing the user's request. For reasons I'll explain, I was hoping to avoid linking against cuBLAS here, but no problem for now.

File: tests/symmetric_contraction_test.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+
+@pytest.fixture(params=[torch.float32, torch.float64], ids=["F32", "F64"])
+def dtype(request):
+    return request.param
+
+
+@pytest.fixture
+def device():
+    if not torch.cuda.is_available():
```
Comment (vbharadwaj-bk) [inline tests/symmetric_contraction_test.py]: No need for this to be a fixture. Tests should fail if there is no torch.device("cuda"), not skip. I would use torch.device("cuda") inline everywhere.

Comment (vbharadwaj-bk) [reply tests/symmetric_contraction_test.py]: I would just declare a global

`DEVICE=torch.device("cuda")`

File: tests/symmetric_contraction_test.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        )
+    if not oeq.BUILT_EXTENSION:
+        pytest.skip(
+            f"OpenEquivariance extension is not built: {oeq.BUILT_EXTENSION_ERROR}"
+        )
+    return torch.device("cuda")
+
+
+@pytest.fixture
+def labels(device):
```
Comment (vbharadwaj-bk) [inline tests/symmetric_contraction_test.py]: Ditto, don't need this as a fixture.

File: tests/symmetric_contraction_test.py [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+        and target_params[name].requires_grad
+        and target_params[name].shape == param.shape
+    ]
+    assert names, "No matching trainable parameters found"
+    return tuple(source_params[name] for name in names), tuple(
+        target_params[name] for name in names
+    )
+
+
+def random_like(tensor, seed):
```
Comment (vbharadwaj-bk) [inline tests/symmetric_contraction_test.py:113]: Hmm.. I would set the generator state outside this function. Otherwise every tensor of the same size gets the same random values (?)

**Review by vbharadwaj-bk — CHANGES_REQUESTED** [review]
Overall, looks great. Couple of things:

1. RocM, I'm assuming hasn't been tested. Are you basically adding back in code that was taken out? In which case, no problem. Otherwise, I'm a bit wary.
2. Have you run the test suite / installation on google collab with an NVIDIA GPU, or on a platform of your choice? I would do that, since this requires changes to the low-level extension.
3. No problem adding cuBLAS back to the CMakeLists.txt for the stable extension, since someone wants this. I was really hoping to not do this; without it, we could package up a prebuilt wheel and avoid compilation on the user side entirely. Now, we can't do that, and worse, the user's cuBLAS should be ABI compatible with PyTorch's installed version. At some point, we should address this.
4. Future work (not in this commit): elimination of the groupMM custom class. You can switch to a caching scheme similar to the regular OEQ kernels for the cuBLASLt state. This will enable you to compose this with `torch.compile` (and, incidentally, you should tell the user that this is not compatible with torch.compile (yet). Then you can add a symmetric contraction test in the export_test.py file.

File: openequivariance/CMakeLists.txt (hunk 4 above)
Comment (asglover) [reply openequivariance/CMakeLists.txt:162]: No, this hasn't been tested, I don't have any AMD access through work. I pinged again (one year later) AMD about CI resources and haven't heard back. Do you have some AMD access? or have ideas on how to get some? I'll create a dev account to get some free credits and will use that going forward.

File: openequivariance/CMakeLists.txt (hunk 5 above)
Comment (asglover) [reply openequivariance/CMakeLists.txt:141]: The thought of pre-compiling and distributing a wheel is a good one. 
Eventually PyTorch will get good dtype and backend support for their GroupedMM operator and we can stop vendoring this wrapper.

File: openequivariance/CMakeLists.txt (hunk 3 above)
Comment (asglover) [reply openequivariance/CMakeLists.txt:120]: It needs to be checked, it was never done through CMAKE only through the python extension.

File: openequivariance/openequivariance/_torch/extlib/__init__.py (hunk 1 above)
Comment (asglover) [reply openequivariance/openequivariance/_torch/extlib/__init__.py]: This hasn't been tested. I'm looking though and I'm not sure that we actually linked it in the past somehow? We just referred to "rocblas/rocblas.h" and maybe hiprtc found this header. But I think we'll have to link here for the stable ABI. I'll setup a test on AMD's cloud.

File: tests/symmetric_contraction_test.py (hunk 8 above)
Comment (asglover) [reply tests/symmetric_contraction_test.py:113]: I think this function returns a tensor with values that are determined by the seed, which is passed into this function, so you would only get tensors with the same values in the same places if you called this function with the same seed. The generator should be the pseudorandom function, which can be the same, and is often the same. I would only switch the generator if it was a very expensive method or something.

Comment (asglover) [reply tests/symmetric_contraction_test.py:113]: The way that it hardcodes the random seeds through does bother me. I'll look into if there are good practices about getting the random seed from pytest or something. So it can be repeatable when it needs to be, but otherwise is random.

File: tests/symmetric_contraction_test.py (hunk 7 above)
Comment (asglover) [reply tests/symmetric_contraction_test.py]: I kind of like using a fixture so it makes a new tensor for each unique tests so nothing strange happens with lifetimes or gradients.

File: tests/symmetric_contraction_test.py (hunk 8 above)
Comment (asglover) [reply tests/symmetric_contraction_test.py:113]: I want to add hypothesis for property testing eventually, so I'll delay getting fancy with the random seeds here.

File: openequivariance/CMakeLists.txt (hunk 4 above)
Comment (vbharadwaj-bk) [reply openequivariance/CMakeLists.txt:162]: Ok.. let's leave this in. The pattern looks fine, if someone complains, fixable.

File: tests/symmetric_contraction_test.py (hunk 8 above)
Comment (vbharadwaj-bk) [reply tests/symmetric_contraction_test.py:113]: Yeah I think you can achieve this just by avoiding passing the seed at all as a fixture and initializing the generator once as a global variable. That way, it's repeatable and you get fresh randomness for every tensor that calls the generator.

Comment (asglover) [reply tests/symmetric_contraction_test.py:113]: That would make it test order / number of tests dependent. Running more tests would change the seed. I feel like the feature set provided by hypothesis, where they randomly fuzz the seed every time, except when it fails, then they store that and add it to a list which is always replayed is a good design. But probably should be applied evenly. For this PR, I can change it for this to whatever you'd like. I can remove the seeds and just make it a global.

File: openequivariance/openequivariance/_torch/extlib/__init__.py (hunk 1 above)
Comment (asglover) [reply openequivariance/openequivariance/_torch/extlib/__init__.py]: still waiting for AMD cloud

Comment (vbharadwaj-bk) [reply openequivariance/openequivariance/_torch/extlib/__init__.py]: Made some minor updates and tested - but it's good to go

**Review by vbharadwaj-bk — APPROVED** [review]
